apps/web-client/src/lib/__testData__/debugHandlers.ts

```diff
@@ -0,0 +1,226 @@
+
+/**
+* Upserts multiple customers into the database in a single transaction.
+* Each customer must have an id and displayId.
+* Updates the customers' lastModified timestamp automatically.
+*
+* @param {DB} db - Database connection
+* @param {Array<Omit<Customer, "updatedAt">>} customers - Array of customer
+data objects
+* @returns {Promise<{success: boolean, inserted: number, updated: number,
+errors: Array<{id: string, error: string}>}>}
+* @throws {Error} If transaction fails
+*/
+export async function upsertCustomers(db: DB, customers: Array<Omit<Customer, "updatedAt">>): Promise<void> {
+	if (!customers || !Array.isArray(customers) || customers.length === 0) {
+		return;
+	}
+
+	const errors: Array<{ id: number; error: string }> = [];
+
+	return await db.tx(async (txDb) => {
+		for (const customer of customers) {
+			// Validate required fields
+			if (!customer.id) {
+				errors.push({ id: parseInt(customer.displayId), error: "Customer must have an id" });
+				continue;
+			}
+
+			if (!customer.displayId) {
+				errors.push({ id: customer.id, error: "Customer must have a displayId" });
+				continue;
+			}
+
+			await upsertCustomer(txDb, customer);
+		}
+	});
+}
+
+/**
+ * Creates or updates multiple book records in a single transaction.
+ * Uses ISBN as the unique identifier for upsert operations.
+ * All fields except ISBN are optional and will only be updated if provided.
+ *
+ * @param {DB} db - Database connection
+ * @param {BookData[]} books - Array of book metadata objects
+ * @throws {Error} If any book is missing an ISBN
+ * @returns {Promise<void>}
+ */
+export async function upsertBooks(db: DB, books: BookData[]): Promise<void> {
+	if (!books.length) {
+		return;
+	}
+
+	// Validate all books have ISBNs before starting transaction
+	for (const book of books) {
+		if (!book.isbn) {
+			throw new Error("All books must have an ISBN");
+		}
+	}
+
+	return db.tx(async (tx) => {
+		for (const book of books) {
+			await upsertBook(tx, book);
+		}
+	});
+}
+
+/**
+ * Updates existing suppliers or inserts new ones if they don't exist.
+ * Processes multiple suppliers in a single transaction for better performance
+ *
+ * @param db - The database instance to query
+ * @param suppliers - Array of supplier data to upsert
+ * @throws {Error} If any supplier is missing an id
+ * @returns {Promise<void>}
+ */
+export async function upsertSuppliers(db: DB, suppliers: Supplier[]): Promise<void> {
+	if (!suppliers.length) {
+		return;
+	}
+
+	// Validate all suppliers have IDs before starting transaction
+	for (const supplier of suppliers) {
+		if (!supplier.id) {
+			throw new Error("All suppliers must have an id");
+		}
+	}
+
+	return db.tx(async (txDb) => {
+		for (const supplier of suppliers) {
+			await txDb.exec(
+				`INSERT INTO supplier (id, name, email, address)
+VALUES (?, ?, ?, ?)
+ON CONFLICT(id) DO UPDATE SET
+name = COALESCE(?, name),
+email = COALESCE(?, email),
+address = COALESCE(?, address);`,
```

Comment:
The missing data (or non-existence of it) should be handled in data setup (definition), rather than within the function.

This function uses raw SQL, this is not much different from using a full SQL file for db seed (which is the functionality we're replacing here)